VSR: Grid repair protocol

sentientwaffle opened the pull request:
Repair corrupt grid blocks.

(Note that grid corruption is still disabled in the VOPR, pending state sync, because grid repair alone may stall if the block is unavailable. However, I have tested the VOPR locally to check that there are no VOPR issues besides the liveness issue.)

The `Replica.open()` sequence changes somewhat, because grid repair must run during `StateMachine.open()`. So where before the replica could run `StateMachine.open()` "synchronously", it must now run asynchronously. Commits are blocked until this finishes.

Related: a solo replica that starts with `op == commit_min` will now transition immediately to `status=normal` rather than waiting for `view_durable_update()` to complete. And `on_request` will reject incoming requests until `view_durable_update()` completes, to guard against the solo replica committing a message, then restarting and discarding it because it is part of a newer view.

## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance.

sentientwaffle committed:
VSR: Grid repair protocol

Repair corrupt grid blocks.

(Note that grid corruption is still disabled in the VOPR, pending state sync, because grid repair alone may stall if the block is unavailable.
However, I have tested the VOPR locally to check that there are no VOPR issues besides the liveness issue.)

The `Replica.open()` sequence changes somewhat, because grid repair must run during `StateMachine.open()`.
So where before the replica could run `StateMachine.open()` "synchronously", it must now run asynchronously.
Commits are blocked until this finishes.

Related: a solo replica that starts with `op == commit_min` will now transition immediately to `status=normal` rather than waiting for `view_durable_update()` to complete.
And `on_request` will reject incoming requests until `view_durable_update()` completes, to guard against the solo replica committing a message, then restarting and discarding it because it is part of a newer view.

matklad commented on src/vsr/replica.zig:
```diff
@@ -666,6 +719,12 @@ pub fn ReplicaType(
             });
             errdefer self.grid.deinit(allocator);
 
+            for (self.grid_write_blocks) |*block, i| {
+                errdefer for (self.grid_write_blocks[0..i]) |b| allocator.free(b);
+                block.* = try alloc_block(allocator);
+            }
+            errdefer for (self.grid_write_blocks) |b| allocator.free(b);
```
I wonder if it makes sense to have

```zig
self.grid_write_blocks = alloc_blocks(allocator, constants.grid_repair_writes_max);
defere free_blocks(allocator, self.grid_write_blocks);
```

if only to avoid partial errdefers ?

I guess we can't easily express it in zig? We need "return array if size is comptime known, and a slice otherwise".

sentientwaffle replied:
We could make it `alloc_blocks(slice_of_blocks);` (`!void` return type), which would be compatible with arrays or slices (caller must heap-allocate the slice for non-comptime-known capacity). But I think this is a common enough Zig idiom that single allocation is ok, even if it is ugly.

matklad commented on src/vsr/replica.zig:
```diff
@@ -1939,6 +2008,203 @@ pub fn ReplicaType(
+                };
+
+                self.grid.read_block_repair(
+                    on_block_read_repair,
```
Naming convention wise, shouldn't this be `on_request_blocks_read` (on_request_prepare / on_request_prepare_read)?

matklad commented on src/vsr/replica.zig:
```diff
@@ -1939,6 +2008,203 @@ pub fn ReplicaType(
+        }
+
+        fn on_block(self: *Self, message: *const Message) void {
+            maybe(!self.state_machine_opened);
```
```suggestion
            maybe(self.state_machine_opened);
```

I feel like semantically maybe already hints at both possibilities.

matklad commented on src/vsr/replica.zig:
```diff
@@ -1939,6 +2008,203 @@ pub fn ReplicaType(
+        fn on_block(self: *Self, message: *const Message) void {
+            maybe(!self.state_machine_opened);
+            assert(message.header.command == .block);
+            assert(message.header.size <= constants.block_size);
```
Should this be `assert(message.header.size <= constants.block_size + @sizeOf(Header));`?

matklad replied:
Oh, no, it actually is correct, a block is a message! :exploding_head: we set `.command = .block,` in `lsm/table.zig`

sentientwaffle replied:
Yeah, I was on the fence about "wrapping" the block in an additional header when converting it to a message, but there wasn't a strong reason to.

matklad commented on src/vsr/replica.zig:
```diff
@@ -1939,6 +2008,203 @@ pub fn ReplicaType(
+            };
+
+            const block = &self.grid_write_blocks[self.grid_writes.index(write)];
+            stdx.copy_disjoint(.inexact, u8, block.*, message.buffer[0..message.header.size]);
```
:thinking: so we have a memcpy on both read and write path here. This feels like a relatively big deal, as block repair is our most perf-sensitive operation (when we repair, we might have a lot of data to transfer). Why do we not use our pre-allocated message pool here?

matklad commented on src/vsr/replica.zig:
```diff
@@ -1939,6 +2008,203 @@ pub fn ReplicaType(
+            );
+        }
+
+        fn on_block_read_repair(grid_read: *Grid.Read, block_: ?Grid.BlockPtrConst) void {
```
Move that next to `on_request_blocks`, so that they can be read sequentially, as if we had async/await.

matklad commented on src/vsr/replica.zig:
```diff
@@ -2181,6 +2447,57 @@ pub fn ReplicaType(
+                .cluster = self.cluster,
+                .replica = self.replica,
+                .view = self.view,
+                .size = @sizeOf(Header) + @intCast(u32, requests_count) * @sizeOf(vsr.BlockRequest),
```
Maybe `var requests_count: u32 = 0;` instead?

matklad commented on src/vsr/replica.zig:
```diff
@@ -6647,6 +6985,15 @@ pub fn ReplicaType(
+                assert(self.grid.read_faulty_queue.count == 1);
+
+                self.grid_repair_message_timeout.start();
+                self.on_grid_repair_message_timeout();
```
Hm, invoking the timeout manually is not something we do elsewhere, it is a bit surprising. 

Perhaps we should wait until the next replica tick? Right now we'll immediately send a request for a single block, but it doesn't seem unreasonable that more than one block would fail at the same time?

matklad reviewed (commented):
Two non-trivial questions:

* can we avoid memcpys here? feels relatively important
* manually invoking `on_grid_repair_message_timeout` feels iffy. Like, I don't think there are any specific problems there, but I don't like how we can reach that function both from replica.tick, and from grid's callback.

sentientwaffle committed:
VSR: send_request_blocks on next tick

To ensure multiple simultaneous broken blocks are requested once.

sentientwaffle committed:
VSR/Grid: Avoid extra block copies

When handling `request_blocks`, read directly from storage into the message buffer (when possible).
A copy is still required when the block is already in the block cache.

kprotty commented on src/lsm/grid.zig:
```diff
@@ -768,33 +707,102 @@ pub fn GridType(comptime Storage: type) type {
+            read.callback(read, block);
+        }
+
+        /// If the block is not present (or corrupt), we do not attempt to repair it — the repair's
+        /// address/checksum is not assumed to match what "should" be in our block.
+        /// Relatedly we still attempt to read free blocks — they might still hold the requested data.
```
Is there a semantic difference between "not present" and "free" (does the former refer to the cache)?

sentientwaffle replied:
"Not present" means that we don't have the block in our cache _or_ disk.
"Free" means that the block's address is not held by the freeset.

matklad commented on src/vsr/replica.zig:
```diff
@@ -6992,8 +6952,71 @@ pub fn ReplicaType(
+                self.grid.on_next_tick(
+                    on_grid_read_fault_next_tick,
+                    &self.grid_read_fault_next_tick,
+                );
```
Hm, this:

* requests blocks immediately
* and also schedules requesting blocks on the next tick

We should do only the latter, right?

sentientwaffle replied:
Good catch; I left the `send_request_blocks` call in by mistake. :facepalm:

matklad commented on src/lsm/grid.zig:
```diff
@@ -768,33 +707,102 @@ pub fn GridType(comptime Storage: type) type {
+                // 1. Replica A write block X₁ to address X.
+                // 2. Replica A write block X₂ to address X (write is lost/misdirected).
+                // 3. Replica B requests block X₁ from replica A.
+                // It is safe for A to send back X₁, but A must not allow it to poison its cache.
```
Maybe move that to doc comment for `read_block_repair`, and also mention that caching would be bad from perf perspective because: 

* it will fill cache with non-temporally-local blocks
* it will force an extra memcpy.

matklad commented on src/lsm/grid.zig:
```diff
@@ -768,33 +707,102 @@ pub fn GridType(comptime Storage: type) type {
+                    block_offset(read.address),
+                );
+            }
+        }
```
I guess this could be a tiny bit more optimal and avoid introduction of next_tick delay if we are going to rid from storage anyway? 

```
if in_cache {
    memcpy()
    on_next_tick(|| read.callback())
} else {
    read_sectors
}
```

That is, copy data to message synchroneously, and then I think it would be safe to next-tick?

sentientwaffle committed:
VSR/Grid: Tidy up control flow

eatonphil commented on scripts/.cspell.json:
```diff
@@ -127,6 +127,7 @@
 	"checkpointing",
-	"checkpointed"
+	"checkpointed",
+        "freeset"
```
Nice, glad this was somewhat obvious to edit
